File: server/src/main/com/broll/gainea/server/core/utils/PlayerUtils.java
package com.broll.gainea.server.core.utils;

import com.broll.gainea.server.core.GameContainer;
import com.broll.gainea.server.core.map.Location;
import com.broll.gainea.server.core.objects.MapObject;
import com.broll.gainea.server.core.objects.Soldier;
import com.broll.gainea.server.core.objects.Unit;
import com.broll.gainea.server.core.player.Player;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Objects;
import java.util.Optional;
import java.util.Set;
import java.util.function.Consumer;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public final class PlayerUtils {

    public static void iteratePlayers(GameContainer game, int pauseBetween, Consumer<Player> consumer) {
        int current = game.getCurrentTurn();
        ArrayList<Player> players = new ArrayList<>(game.getActivePlayers());
        for (int i = 0; i < players.size(); i++) {
            int nr = (current + i) % players.size();
            consumer.accept(players.get(nr));
            ProcessingUtils.pause(pauseBetween);
        }
    }

    public static boolean isCommanderAlive(Player player) {
        return getCommander(player).isPresent();
    }

    public static Stream<Player> getOtherPlayers(GameContainer game, Player player) {
        return game.getAllPlayers().stream().filter(it -> it != player);
    }

    public static Optional<Soldier> getCommander(Player player) {
        return player.getUnits().stream().filter(PlayerUtils::isCommander).map(it -> (Soldier) it).findFirst();
    }

    public static boolean isCommander(Unit unit) {
        return unit instanceof Soldier && ((Soldier) unit).isCommander();
    }

    public static List<Unit> getUnits(Player player, Location location) {
        return player.getUnits().stream().filter(it -> it.getLocation() == location).collect(Collectors.toList());
    }

    public static List<Unit> getHostileArmy(Player player, Location location) {
        return location.getInhabitants().stream().filter(inhabitant -> isHostile(player, inhabitant))
                .map(o -> (Unit) o).collect(Collectors.toList());
    }

    private static boolean isHostile(Player player, MapObject object) {
        if (object instanceof Unit) {
            return player.getFraction().isHostile((Unit) object);
        }
        return false;
    }

    public static boolean hasHostileArmy(Player player, Location location) {
        return location.getInhabitants().stream().anyMatch(inhabitant -> isHostile(player, inhabitant));
    }

    public static Set<Location> getHostileLocations(GameContainer game, Player player) {
        Set<Location> locations = new HashSet<>();
        getOtherPlayers(game, player).map(Player::getControlledLocations).forEach(locations::addAll);
        locations.removeIf(location -> getHostileArmy(player, location).isEmpty());
        return locations;
    }

    public static Player getOwner(List<Unit> units) {
        return units.stream().map(Unit::getOwner).filter(Objects::nonNull).findFirst().orElse(null);
    }


}

File: server/src/main/com/broll/gainea/server/core/utils/PlayerUtils.kt
package com.broll.gainea.server.core.utils

import com.broll.gainea.server.core.Game
import com.broll.gainea.server.core.map.Location
import com.broll.gainea.server.core.objects.IUnit
import com.broll.gainea.server.core.objects.MapObject
import com.broll.gainea.server.core.objects.Soldier
import com.broll.gainea.server.core.objects.Unit
import com.broll.gainea.server.core.objects.resolve
import com.broll.gainea.server.core.player.Player
import com.broll.gainea.server.core.player.isNeutral

fun Game.getWeakestPlayer() =
        activePlayers.minBy { it.goalHandler.score * 15 + it.units.sumOf { unit -> unit.rank() } }

fun Game.iteratePlayers(pauseBetween: Int, consumer: (Player) -> kotlin.Unit) {
    val current = currentTurn
    val players = ArrayList(activePlayers)
    for (i in players.indices) {
        val nr = (current + i) % players.size
        consumer(players[nr])
        ProcessingUtils.pause(pauseBetween)
    }
}

fun Player.isCommanderAlive() = getCommander() != null

fun Game.getOtherPlayers(player: Player) = allPlayers.filter { it !== player }

fun Player.getCommander() = units.filterIsInstance(Soldier::class.java).find { it.isCommander }

fun IUnit.isCommander() = resolve().run { this is Soldier && this.isCommander }

fun Player.getUnits(location: Location) =
        location.units.filter { it.owner == this }

fun Player.getHostileArmy(location: Location) =
        location.units.filter { isHostile(it) }

fun Player.isHostile(obj: MapObject) =
        if (obj is Unit) {
            this.fraction.isHostile(obj)
        } else false


fun Player.hasHostileArmy(location: Location) = getHostileArmy(location).isNotEmpty()

fun Game.getEnemyLocations(player: Player) =
        getOtherPlayers(player).flatMap { it.controlledLocations }.filter { player.hasHostileArmy(it) }

fun List<Unit>.owner() = find { !it.owner.isNeutral() }?.owner ?: first().owner

